src/plugins/kdb/ldap/libkdb_ldap/ldap_principal2.c

```diff
@@ -446,43 +473,111 @@ krb5_encode_krbsecretkey(krb5_key_data *key_data_in, int n_key_data,
+
+cleanup:
+    free(code);
+    free(key_data);
+    free(p_ret);
+    return err;
+}
+
+/* Decoding ASN.1 encoded key */
+static struct berval **
+krb5_encode_krbsecretkey(krb5_key_data *key_data, int n_key_data,
+                         krb5_kvno mkvno)
+{
+    struct berval **ret = NULL;
+    int     currkvno;
+    int     num_versions = 0;
+    int     i, j, last;
+    krb5_error_code err = 0;
+
+    if (n_key_data < 0)
+        return NULL;
+
     /* Find the number of key versions */
-    for (i = 0; i < n_key_data - 1; i++)
-        if (key_data[i].key_data_kvno != key_data[i + 1].key_data_kvno)
-            num_versions++;
+    if (n_key_data > 0) {
+        for (i = 0, num_versions = 1; i < n_key_data - 1; i++) {
+            if (key_data[i].key_data_kvno != key_data[i + 1].key_data_kvno)
+                num_versions++;
+        }
+    }
 
-    ret = (struct berval **) calloc (num_versions + 1, sizeof (struct berval *));
+    ret = calloc(num_versions + 1, sizeof(struct berval *));
     if (ret == NULL) {
         err = ENOMEM;
         goto cleanup;
     }
-    for (i = 0, last = 0, j = 0, currkvno = key_data[0].key_data_kvno; i < n_key_data; i++) {
-        krb5_data *code;
-        if (i == n_key_data - 1 || key_data[i + 1].key_data_kvno != currkvno) {
-            ret[j] = k5alloc(sizeof(struct berval), &err);
-            if (ret[j] == NULL)
-                goto cleanup;
-            err = asn1_encode_sequence_of_keys(key_data + last,
-                                               (krb5_int16)i - last + 1,
-                                               mkvno, &code);
-            if (err)
-                goto cleanup;
-            /*CHECK_NULL(ret[j]); */
-            ret[j]->bv_len = code->length;
-            ret[j]->bv_val = code->data;
-            free(code);
-            j++;
-            last = i + 1;
-
-            if (i < n_key_data - 1)
-                currkvno = key_data[i + 1].key_data_kvno;
+    ret[num_versions] = NULL;
+
+    /* n_key_data may be 0 if a policy is created without a key. */
+    if (n_key_data > 0) {
+        for (i = 0, last = 0, j = 0, currkvno = key_data[0].key_data_kvno;
+             i < n_key_data; i++) {
+            if (i == n_key_data - 1 ||
+                key_data[i + 1].key_data_kvno != currkvno) {
+                err = encode_keys(key_data + last, (krb5_int16)i - last + 1,
+                                  mkvno, ret + j);
+                if (err)
+                    goto cleanup;
+                j++;
+                last = i + 1;
+
+                if (i < n_key_data - 1)
+                    currkvno = key_data[i + 1].key_data_kvno;
+            }
         }
     }
-    ret[num_versions] = NULL;
 
 cleanup:
+    if (err != 0) {
+        free_berdata(ret);
+        ret = NULL;
+    }
 
-    free(key_data);
+    return ret;
+}
+
+/*
+ * Encode a principal's key history for insertion into ldap.
+ */
+static struct berval **
+krb5_encode_histkey(osa_princ_ent_rec *princ_ent) {
```

Comment:
Open brace goes on the next line.  (This was in Tomas's PR but still needs to be fixed.)